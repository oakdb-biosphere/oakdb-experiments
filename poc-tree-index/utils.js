const { dset } = require("dset");

async function series(...args) {
  for (arg in args) {
    await arg;
  }
}

function parallel(...args) {
  return Promise.all(args);
}

function uuid() {
  return Math.random().toString(16).substring(2);
}

function zip(list) {
  return list.reduce((a, { k, v }) => ((a[k] = v), a), {});
}

function collapse(list) {
  const obj = {};
  for (const { k, v } of list) {
    dset(obj, k.replaceAll("/", "."), v);
  }
  return obj;
}

exports.series = series;
exports.parallel = parallel;
exports.uuid = uuid;
exports.zip = zip;
exports.collapse = collapse;
