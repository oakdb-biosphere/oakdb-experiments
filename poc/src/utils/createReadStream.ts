import { dset } from "dset";

// dset: A tiny (194B) utility for safely writing deep Object values
// const dset = (obj, a, b) => (obj[a] = b);

export function createReadStreamAsObj<T>(
  db,
  from: string | number,
  to: string | number,
  { sep = ":" }
): Promise<Record<string, T>> {
  const obj: Record<string, T> = {};
  return new Promise((resolve, reject) => {
    db.createReadStream({
      gte: from,
      lt: to,
    })
      .on("data", (data: { key: string; value: any }) =>
        dset(
          obj,
          data.key.replace(from + sep, "").replaceAll(/[\/:]/g, "."),
          data.value
        )
      )
      .on("error", reject)
      .on("close", () => resolve(obj));
  });
}

export function createReadStreamAsArray<T>(
  db,
  from: string | number,
  to: string | number
): Promise<T[]> {
  const obj: T[] = [];
  return new Promise((resolve, reject) => {
    db.createReadStream({
      gte: from,
      lt: to,
    })
      .on("data", (data: T) => obj.push(data))
      .on("error", reject)
      .on("close", () => resolve(obj));
  });
}

export default createReadStreamAsObj;
