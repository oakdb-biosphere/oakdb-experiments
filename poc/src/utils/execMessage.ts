export async function execMessage(db, msg, { $log }) {
  let result = null;
  $log("‚úâÔ∏è", msg);

  switch (msg.op) {
    case "init": {
      await db.createObject("/articles", "id1", { foo: "bar", baz: "oak" });
      await db.createObject("/articles", "id2", { foo: "qux", baz: "db" });
      await db.createObject("/articles/id1/sub", "id3", { hello: "world" });
      break;
    }

    case "create": {
      if (typeof msg.value !== "object") {
        throw new Error("value must be an object");
      }
      await db.createObject(msg.path, msg.id, msg.value);
      break;
    }

    case "update":
    case "set": {
      if (msg.prop) {
        // msg.path?
        if (msg.pathNotFound) {
          $log("üîµ", "createObject due to not found", { ...msg });
          await db.createObject(msg.pathAssumingDoc, msg.id, {
            [msg.prop]: msg.value,
          });
        } else {
          await db.setProp(msg.id, msg.prop, msg.value);
        }
      } else if (msg.value === null) {
        await db.deleteObjectByPath(msg.path + "/" + msg.id);
        // await db.deleteObjectById(msg.id);
      } else {
        if (msg.pathNotFound) {
          await db.createObject(msg.path, msg.id, msg.value);
        } else if (msg.op === "update") {
          await db.updateObject(msg.id, msg.value);
        } else {
          await db.createObject(msg.path, msg.id, msg.value);
        }
      }
      break;
    }

    case "delete": {
      if (msg.pathNotFound) {
        await db.deleteObjectByPath(msg.path + "/" + msg.id);
      } else {
        await db.deleteObjectById(msg.id);
      }
      break;
    }

    case "subscribe":
    case "query": {
      if (msg.prop) {
        result = await db.getProp(msg.id, msg.prop);
      } else if (msg.pathNotFound) {
        result = await db.getObjectsByPath(msg.path);
      } else {
        result = await db.getObjectById(msg.id);
      }
      break;
    }

    case "clear": {
      await db.clear();
      break;
    }
  }

  return result;
}
