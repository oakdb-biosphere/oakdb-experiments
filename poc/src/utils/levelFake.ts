function stream(
  db: Map<string, any>,
  { gte, lt }: { gte?: string; lt?: string } = {}
) {
  const callbacks = {
    data: (obj: any) => {},
    error: () => {},
    close: () => {},
    end: () => {},
  };

  setTimeout(() => {
    for (const key of db.keys()) {
      if (key.localeCompare(gte) >= 0 && key.localeCompare(lt) < 0) {
        callbacks?.data({ key, value: db.get(key) });
      }
    }
    callbacks?.close();
    callbacks?.end();
  });

  return {
    on(event, cb) {
      callbacks[event] = cb;
      return this;
    },
  };
}

type TAllowed = string | number | boolean;

function level(path, options) {
  const db = new Map<string, TAllowed>();

  return {
    async put(key: string, value: TAllowed) {
      db.set(key, value);
    },
    async del(key: string) {
      db.delete(key);
    },
    async get(key: string) {
      return db.get(key);
    },
    createReadStream(options) {
      return stream(db, options);
    },
    async clear() {
      db.clear();
    },
    async open() {},
    async close() {},
    isOpen: () => true,
    isClosed: () => true,
    isOperational: () => true,
    // getMany: InferDBGetMany<DB>;
    $debug() {
      return JSON.parse(JSON.stringify([...db.entries()], null, 2));
    },
  };
}

export default level;
