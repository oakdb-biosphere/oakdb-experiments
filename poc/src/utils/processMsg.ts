export async function processMsg(db, msgOriginal, { $log }) {
  const msg = { ...msgOriginal };
  msg.pathBackup = msg.path;

  if (msg.path?.includes(":")) {
    const [path, prop] = msg.path.split(":");

    if (msg.prop && msg.prop !== prop) {
      throw new Error("Prop provided twice");
    }

    msg.path = path;
    msg.prop = prop;
  }

  if (msg.id?.includes(":")) {
    const [id, prop] = msg.id.split(":");

    if (msg.prop && msg.prop !== prop) {
      throw new Error("Prop provided twice");
    }

    msg.id = id;
    msg.prop = prop;
  }

  if (msg.path && !msg.id) {
    const segments = msg.path.split("/");
    msg.id = segments[segments.length - 1];
    msg.path = segments.slice(0, -1).join("/");
  }

  msg.pathAssumingDoc = msg.path;
  msg.path = await db.getPathById(msg.id);

  if (!msg.path) {
    // $log("msg", JSON.stringify(msg, null, 2));
    msg.path = msg.pathBackup;
    msg.pathNotFound = true;
  }

  if (msg.path) {
    // Remove double slashes
    msg.path = msg.path.replace(/(\/)+/g, "$1");
  }

  if (msg.path && msg.path.endsWith("/")) {
    msg.path = msg.path.substring(0, msg.path.length - 1);
  }

  delete msg.pathBackup;

  // TODO:
  msg.connectionId = "con:1";

  // TODO: Aliases?
  return msg;
}
