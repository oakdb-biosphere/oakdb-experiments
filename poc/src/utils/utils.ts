import { dset } from "dset";

export async function series(...args: (Promise<any> | any)[]) {
  for (const arg in args) {
    await arg;
  }
}

export function parallel(...args: (Promise<any> | any)[]) {
  return Promise.all(args);
}

export function uuid() {
  return Math.random().toString(16).substring(2);
}

export function zip(list) {
  return list.reduce((a, { k, v }) => ((a[k] = v), a), {});
}

export function collapse(list: any[]): any {
  const obj = {};
  for (const { k, v } of list) {
    dset(obj, k.replaceAll("/", "."), v);
  }
  return obj;
}
