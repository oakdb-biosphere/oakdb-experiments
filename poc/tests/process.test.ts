import { processMsg } from "../src/utils/processMsg";
import messages from "./__fixture__/messages";

describe("processMsg", () => {
  const mockDb = {
    getPathById(id) {
      if (id === "uuid" || id === "<uuid>" || id === "foo")
        return "/path/to/docs/";
      else return null;
    },
  };

  const messagesAfterTransformation = [];

  const title = (msg, index) =>
    [
      `[${index}]`,
      `${msg.op}`,
      `${msg.path || msg.id || "<no-path-or-id>"}:${msg.prop || "*"}`,
    ].join(" ");

  describe("Messages", () => {
    messages.forEach((msg, index) => {
      it(title(msg, index), async () => {
        const messageTransformed = await processMsg(mockDb, msg, {
          $log: () => {},
        });
        expect(typeof messageTransformed).toBe("object");
        expect(messageTransformed.op).toBe(msg.op);
        if (!messageTransformed.pathNotFound) {
          expect(messageTransformed.path).toBeDefined();
          expect(typeof messageTransformed.path).toBe("string");
        }
        expect(messageTransformed.path).not.toMatch(/(\/){2,}/g);
        expect(messageTransformed.id).toBeDefined();
        messagesAfterTransformation.push(messageTransformed);
      });
    });
  });

  it("Check all (snapshot)", () => {
    // console.log(JSON.stringify(messagesAfterTransformation, null, 2));
    expect(messagesAfterTransformation).toMatchSnapshot();
  });
});
