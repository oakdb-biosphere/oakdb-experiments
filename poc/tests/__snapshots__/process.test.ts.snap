// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processMsg Check all (snapshot) 1`] = `
Array [
  Object {
    "id": "<uuid>",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "<uuid>",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "value": Object {
      "obj": "boo",
      "v": "foo",
    },
  },
  Object {
    "id": "<uuid>",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": undefined,
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "<uuid>",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": undefined,
    "value": Object {
      "obj": "boo",
      "v": "foo",
    },
  },
  Object {
    "id": "foo",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "docs",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to",
    "pathNotFound": true,
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "foo",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "foo",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
    "value": null,
  },
  Object {
    "id": "foo",
    "op": "set",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "value": null,
  },
  Object {
    "id": "to",
    "op": "set",
    "path": "/path/to",
    "pathAssumingDoc": "/path",
    "pathNotFound": true,
    "value": null,
  },
  Object {
    "id": "<uuid>",
    "op": "query",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
  },
  Object {
    "id": "<uuid>",
    "op": "query",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
  },
  Object {
    "id": "docs",
    "op": "query",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to",
    "pathNotFound": true,
  },
  Object {
    "id": "path",
    "op": "query",
    "path": "/path",
    "pathAssumingDoc": "",
    "pathNotFound": true,
  },
  Object {
    "id": "<uuid>",
    "op": "subscribe",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
  },
  Object {
    "id": "<uuid>",
    "op": "subscribe",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
  },
  Object {
    "id": "docs",
    "op": "subscribe",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to",
    "pathNotFound": true,
  },
  Object {
    "id": "path",
    "op": "subscribe",
    "path": "/path",
    "pathAssumingDoc": "",
    "pathNotFound": true,
  },
  Object {
    "id": "<uuid>",
    "op": "create",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "value": Object {},
  },
  Object {
    "id": "<uuid>",
    "op": "create",
    "path": "/path/to/docs",
    "pathAssumingDoc": undefined,
    "value": Object {},
  },
  Object {
    "id": "<uuid>",
    "op": "update",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "<uuid>",
    "op": "update",
    "path": "/path/to/docs",
    "pathAssumingDoc": undefined,
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "<uuid>",
    "op": "update",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "<uuid>",
    "op": "update",
    "path": "/path/to/docs",
    "pathAssumingDoc": undefined,
    "prop": "text",
    "value": "boo",
  },
  Object {
    "id": "to",
    "op": "delete",
    "path": "/path/to",
    "pathAssumingDoc": "/path",
    "pathNotFound": true,
  },
  Object {
    "id": "<uuid>",
    "op": "query",
    "path": "/path/to/docs",
    "pathAssumingDoc": "/path/to/docs",
    "prop": "text",
  },
]
`;
